openvdb/openvdb/tools/Merge.h

```diff
@@ -0,0 +1,931 @@
+
+#ifndef OPENVDB_TOOLS_MERGE_HAS_BEEN_INCLUDED
+#define OPENVDB_TOOLS_MERGE_HAS_BEEN_INCLUDED
+
+#include <openvdb/Platform.h>
+#include <openvdb/Exceptions.h>
+#include <openvdb/Types.h>
+#include <openvdb/Grid.h>
+#include <openvdb/tree/NodeManager.h>
+
+#include <unordered_map>
+#include <unordered_set>
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace tools {
+
+
+/// @brief Convenience class that contains a pointer to a const or a non-const tree
+/// and a subset of methods to retrieve data from the tree.
+///
+/// @details This class has methods to extract data from the tree that will either
+/// result in stealing the data or deep-copying it depending on whether the
+/// DeepCopy or Steal tag dispatch class was used in its construction.
+///
+/// @details Stealing data requires a non-const tree pointer. There is a constructor
+/// to pass in a tree shared pointer for cases where it is desirable for this class
+/// to maintain shared ownership.
+template <typename TreeT>
+struct TreeToMerge
+{
+    using TreeType = std::remove_const_t<TreeT>;
+    using RootNodeType = typename TreeType::RootNodeType;
+    using ValueType = typename TreeType::ValueType;
+    using MaskTreeType = typename TreeT::template ValueConverter<ValueMask>::Type;
+
+    TreeToMerge() = default;
+
+    /// @brief Non-const pointer tree constructor for stealing data.
+    TreeToMerge(TreeType* tree, Steal)
+        : mTree(tree) { }
+    /// @brief Non-const shared pointer tree constructor for stealing data.
+    TreeToMerge(typename TreeType::Ptr treePtr, Steal)
+        : mTreePtr(treePtr), mTree(mTreePtr.get()) { }
+
+    /// @brief Const tree pointer constructor for deep-copying data. As the
+    /// tree is not mutable and thus cannot be pruned, a lightweight mask tree
+    /// with the same topology is created that can be pruned to use as a
+    /// reference. Initialization of this mask tree can optionally be disabled
+    /// for delayed construction.
+    TreeToMerge(const TreeType* constTree, DeepCopy, bool initialize = true)
+        : mConstTree(constTree)
+    {
+        if (mConstTree && initialize)     this->initializeMask();
+    }
+
+    /// @brief Non-const tree pointer constructor for deep-copying data. The
+    /// tree is not intended to be modified so is not pruned, instead a
+    /// lightweight mask tree with the same topology is created that can be
+    /// pruned to use as a reference. Initialization of this mask tree can
+    /// optionally be disabled for delayed construction.
+    TreeToMerge(TreeType* tree, DeepCopy tag, bool initialize = true)
+        : TreeToMerge(static_cast<const TreeType*>(tree), tag, initialize) { }
+
+    /// @brief Reset the non-const tree shared pointer. This is primarily
+    /// used to preserve the order of trees to merge in a container but have
+    /// the data in the tree be lazily loaded or resampled.
+    void reset(typename TreeType::Ptr treePtr)
```

Comment:
This bypasses the Steal flag?   Should it take the Steal Flag?

Does it make sense to reset() a tree to a DeepCopy mode for the same delay-load argument?

Or is reset() entirely internal to the operators working on TreeToMerge?

Reply:
No, reset() is not entirely internal. I think for consistency it makes sense to add it, especially if we represent the mode by a state flag instead of whether the tree is const/non-const. I don't really see a good use case for this, but perhaps there's one I'm missing. I'll change this.